File: speech_to_text/go/transcribe_file/remote/remote.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"
)

func main() {
	apiBase := "https://api.soniox.com"
	apiKey := os.Getenv("SONIOX_API_KEY")
	audioUrl := "https://soniox.com/media/examples/coffee_shop.mp3"

	// 1. Start a new transcription session by sending the audio URL to the API
	fmt.Println("Starting transcription...")
	transcriptionPayload := map[string]string{
		"audio_url": audioUrl,
		"model":     "stt-async-preview",
	}
	transcriptionBody, err := json.Marshal(transcriptionPayload)
	if err != nil {
		log.Fatalf("Failed to marshal request payload: %v", err)
	}

	req, err := http.NewRequest("POST", fmt.Sprintf("%s/v1/transcriptions", apiBase), bytes.NewReader(transcriptionBody))
	if err != nil {
		log.Fatalf("Failed to create request: %v", err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)
	req.Header.Set("Content-Type", "application/json")

	res, err := http.DefaultClient.Do(req)
	if err != nil {
		log.Fatalf("Request failed: %v", err)
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusCreated {
		log.Fatalf("Request failed with status: %s", res.Status)
	}

	var transcription map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&transcription); err != nil {
		log.Fatalf("Failed to decode response: %v", err)
	}

	transcriptionID, ok := transcription["id"].(string)
	if !ok {
		log.Fatalf("Failed to get transcription ID from response")
	}
	fmt.Println("Transcription started with ID: " + transcriptionID)

	// 2. Poll the transcription endpoint until the status is 'completed'
	for {
		req, err := http.NewRequest("GET", fmt.Sprintf("%s/v1/transcriptions/%s", apiBase, transcriptionID), nil)
		if err != nil {
			log.Fatalf("Failed to create request: %v", err)
		}
		req.Header.Set("Authorization", "Bearer "+apiKey)

		res, err := http.DefaultClient.Do(req)
		if err != nil {
			log.Fatalf("Request failed: %v", err)
		}
		defer res.Body.Close()

		if res.StatusCode != http.StatusOK {
			log.Fatalf("Request failed with status: %s", res.Status)
		}

		var statusResponse map[string]interface{}
		if err := json.NewDecoder(res.Body).Decode(&statusResponse); err != nil {
			log.Fatalf("Failed to decode response: %v", err)
		}

		if status, ok := statusResponse["status"].(string); ok {
			if status == "error" {
				errorMessage, _ := statusResponse["error_message"].(string)
				log.Fatalf("Transcription error: %s", errorMessage)
			} else if status == "completed" {
				break
			}
		}

		// Wait for 1 second before polling again
		time.Sleep(1 * time.Second)
	}

	// 3. Retrieve the final transcript once transcription is completed
	req, err = http.NewRequest("GET", fmt.Sprintf("%s/v1/transcriptions/%s/transcript", apiBase, transcriptionID), nil)
	if err != nil {
		log.Fatalf("Failed to create request: %v", err)
	}
	req.Header.Set("Authorization", "Bearer "+apiKey)

	res, err = http.DefaultClient.Do(req)
	if err != nil {
		log.Fatalf("Request failed: %v", err)
	}
	defer res.Body.Close()

	if res.StatusCode != http.StatusOK {
		log.Fatalf("Request failed with status: %s", res.Status)
	}

	var transcriptResponse map[string]interface{}
	if err := json.NewDecoder(res.Body).Decode(&transcriptResponse); err != nil {
		log.Fatalf("Failed to decode response: %v", err)
	}

	transcriptText, ok := transcriptResponse["text"].(string)
	if !ok {
		log.Fatalf("Failed to get transcript text from response")
	}

	fmt.Printf("Transcript:\n%s\n", transcriptText)
}

File: speech_to_text/ruby/transcribe_file/remote/remote.rb
require 'net/http'
require 'json'
require 'uri'
require 'openssl'

api_base = "https://api.soniox.com"
file_to_transcribe = "https://soniox.com/media/examples/coffee_shop.mp3"

# Retrieve the API key from environment variable (ensure SONIOX_API_KEY is set)
api_key = ENV["SONIOX_API_KEY"]

# Create a requests session and set the Authorization header
http = Net::HTTP
uri = URI(api_base)
headers = {
  "Authorization" => "Bearer #{api_key}"
}

# 1. Start a new transcription session by sending the audio URL to the API
puts "Starting transcription..."
uri = URI("#{api_base}/v1/transcriptions")
request = Net::HTTP::Post.new(uri, headers)
request.body = {
  "audio_url" => file_to_transcribe,
  "model" => "stt-async-preview"
}.to_json
request["Content-Type"] = "application/json"

response = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
  http.request(request)
end

raise "Transcription request failed: #{response.body}" unless response.is_a?(Net::HTTPSuccess)

transcription = JSON.parse(response.body)

transcription_id = transcription["id"]
puts "Transcription started with ID: #{transcription_id}"

# 2. Poll the transcription endpoint until the status is 'completed'
loop do
  uri = URI("#{api_base}/v1/transcriptions/#{transcription_id}")
  request = Net::HTTP::Get.new(uri, headers)

  response = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
    http.request(request)
  end

  raise "Transcription status check failed: #{response.body}" unless response.is_a?(Net::HTTPSuccess)

  transcription = JSON.parse(response.body)

  case transcription["status"]
  when "error"
    raise "Transcription error: #{transcription["error_message"]}"
  when "completed"
    break
  else
    # Wait for 1 second before polling again
    sleep 1
  end
end

# 3. Retrieve the final transcript once transcription is completed
uri = URI("#{api_base}/v1/transcriptions/#{transcription_id}/transcript")
request = Net::HTTP::Get.new(uri, headers)

response = http.start(uri.hostname, uri.port, use_ssl: true) do |http|
  http.request(request)
end

raise "Transcript fetch failed: #{response.body}" unless response.is_a?(Net::HTTPSuccess)

transcript = JSON.parse(response.body)

# Print the transcript text
puts "Transcript:"
puts "#{transcript["text"]}"
